src/wdmp-c.c

```
@@ -47,67 +47,82 @@
 /*                             External Functions                             */
 /*----------------------------------------------------------------------------*/
 
-void wdmp_parse_request(char * payload, req_struct **reqObj)
+void wdmp_parse_generic_request(char * payload, PAYLOAD_TYPE payload_type, void **reqObj)
 {
     cJSON *request = NULL;
     char *out = NULL, *command = NULL;
 
-    request=cJSON_Parse(payload);
+    if (!payload || !reqObj)
+    {
+        printf("wdmp_parse_generic_request - invalid param!\n");
+        return;
+    }
 
-    if(request != NULL)
+    request = cJSON_Parse(payload);
+    if (request != NULL)
     {
         command = cJSON_GetObjectItem(request, "command")->valuestring;
-        WdmpPrint("command %s\n",(command == NULL) ? "NULL" : command);
+        WdmpPrint("command %s\n", (command == NULL) ? "NULL" : command);
 
-        if( command != NULL) 
+        if (command != NULL)
         {
             out = cJSON_PrintUnformatted(request);
-            (*reqObj) = (req_struct *) malloc(sizeof(req_struct));
-            memset( (*reqObj), 0, sizeof( req_struct ) );
 
-            if ((strcmp(command, "GET") == 0)|| (strcmp(command, "GET_ATTRIBUTES") == 0))
+            //allocate structure according to payload type
+            if (payload_type == WDMP_TR181 || payload_type == WDMP_SNMP)
+            {
+                (*reqObj) = (req_struct *) malloc(sizeof(req_struct));
+                memset((*reqObj), 0, sizeof(req_struct));
+            }
+            else
+            {
+                // allocate according to payload type.
```

If not supported, don't silently fail.  Assert/return NULL something.